src/routes/teams.rs

```diff
@@ -1,25 +1,227 @@
+        None,
+        None,
+        Some(true),
+        &**pool,
+    )
+    .await?;
+
+    Ok(HttpResponse::Ok().body(""))
+}
+
+#[post("{id}/members")]
+pub async fn add_team_member(
+    req: HttpRequest,
+    info: web::Path<(TeamId,)>,
+    pool: web::Data<PgPool>,
+    new_member: web::Json<crate::models::teams::TeamMember>,
+) -> Result<HttpResponse, ApiError> {
+    let team_id = info.into_inner().0.into();
+
+    let mut transaction = pool
+        .begin()
+        .await
+        .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+    let current_user = get_user_from_headers(req.headers(), &**pool)
+        .await
+        .map_err(|_| ApiError::AuthenticationError)?;
+    let team_member =
+        TeamMember::get_from_user_id(team_id, current_user.id.into(), &**pool).await?;
+
+    if let Some(member) = team_member {
+        if member.permissions.contains(Permissions::MANAGE_INVITES)
+            && new_member.role != crate::models::teams::OWNER_ROLE
+        {
+            if !member.permissions.contains(new_member.permissions) {
+                return Err(ApiError::AuthenticationError);
+            }
+
+            let new_id =
+                crate::database::models::ids::generate_team_member_id(&mut transaction).await?;
+            TeamMember {
+                id: new_id,
+                team_id,
+                user_id: new_member.user_id.into(),
+                name: new_member.name.clone(),
+                role: new_member.role.clone(),
+                permissions: new_member.permissions,
+                accepted: false,
+            }
+            .insert(&mut transaction)
+            .await
+            .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+            transaction
+                .commit()
+                .await
+                .map_err(|e| ApiError::DatabaseError(e.into()))?;
+
+            Ok(HttpResponse::Ok().body(""))
+        } else {
+            Err(ApiError::AuthenticationError)
+        }
+    } else {
+        Err(ApiError::AuthenticationError)
+    }
+}
+
+#[derive(Serialize, Deserialize, Clone)]
+pub struct EditTeamMember {
+    pub permissions: Option<Permissions>,
+    pub role: Option<String>,
+}
+
+#[patch("{id}/members/{user_id}")]
+pub async fn edit_team_member(
+    req: HttpRequest,
+    info: web::Path<(TeamId, UserId)>,
+    pool: web::Data<PgPool>,
+    edit_member: web::Json<EditTeamMember>,
+) -> Result<HttpResponse, ApiError> {
+    let ids = info.into_inner();
+    let id = ids.0.into();
+    let user_id = ids.1.into();
+
+    let current_user = get_user_from_headers(req.headers(), &**pool)
+        .await
+        .map_err(|_| ApiError::AuthenticationError)?;
+    let team_member = TeamMember::get_from_user_id(id, current_user.id.into(), &**pool).await?;
+
+    if let Some(member) = team_member {
+        if member.permissions.contains(Permissions::EDIT_MEMBER)
+            && edit_member.role.as_deref() != Some(crate::models::teams::OWNER_ROLE)
```

Comment:
There needs to be some way for a user to modify their own name as a team member; it seems like it could fit in this method, but you would have to be careful to make sure that the permissions and role are not changed.

Reply:
No, the team member's name is their username